crates/lib/src/action/workingdata.rs

```diff
@@ -116,31 +120,118 @@ impl Default for WorkingData {
+}
+
+impl std::fmt::Debug for Action {
+	fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+		f.debug_struct("Action")
+			.field("events", &self.events)
+			.field("processes", &self.processes)
+			.finish()
+	}
 }
 
 impl Action {
-	pub(super) fn new(events: Arc<[Event]>) -> Self {
+	pub(crate) fn new(events: Arc<[Event]>, processes: Arc<[SupervisorId]>) -> Self {
 		Self {
 			events,
-			outcome: Default::default(),
+			processes,
+			outcomes: Default::default(),
 		}
 	}
 
-	/// Set the action's outcome.
+	/// Set the action's outcome for all [`Command`]es.
 	///
 	/// This takes `self` and `Action` is not `Clone`, so it's only possible to call it once.
-	/// Regardless, if you _do_ manage to call it twice, it will do nothing beyond the first call.
 	///
 	/// See the [`Action`] documentation about handlers to learn why it's a bad idea to clone or
 	/// send it elsewhere, and what kind of handlers you cannot use.
 	pub fn outcome(self, outcome: Outcome) {
-		self.outcome.set(outcome).ok();
+		let mut outcomes = tokio::task::block_in_place(|| self.outcomes.blocking_lock());
+		for process in self.processes.iter().copied().collect::<Vec<_>>() {
+			self.on_command_with_lock(&mut outcomes, process, outcome.clone(), EventSet::All);
+		}
+	}
+
+	/// Sets an [`Outcome`] for a single [`Command`].
+	///
+	/// The [`EventSet`] provides the specific set of [`Event`]s associated with the [`Command`]
+	/// and [`Outcome`].
+	pub async fn on_command(&self, command: SupervisorId, outcome: Outcome, set: EventSet) {
+		let mut guard = self.outcomes.lock().await;
+		self.on_command_with_lock(&mut guard, command, outcome, set);
 	}
+
+	// Used internally by on_command and outcome to insert into `outcomes`.
+	fn on_command_with_lock(
+		&self,
+		guard: &mut MutexGuard<'_, HashMap<SupervisorId, (Resolution, EventSet)>>,
+		process: SupervisorId,
+		outcome: Outcome,
+		set: EventSet,
+	) {
+		guard.insert(process, (Resolution::Apply(outcome), set));
+	}
+
+	/// Returns a snapshot of the [`SupervisorId`]s of the alive [`Command`]s at creation time of the
+	/// [`Action`].
+	pub fn current_commands(&self) -> &[SupervisorId] {
+		&self.processes
+	}
+
+	/// Starts a new supervised [`Command`].
+	///
+	/// This instantiates a new command supervisor, which manages the lifecycle
+	/// of a command within the watchexec instance, including stopping, starting
+	/// again, sending signals, and monitoring its aliveness.
+	///
+	/// You can control it by calling `command_outcome()` with the [`SupervisorId`]
+	/// this method returns. To destroy a supervisor, use `Outcome::End`.
+	///
+	/// For details on the `events` argument, see the documentation for
+	/// `command_outcome`.
+	///
+	/// Note that as this is async, your action handler must also be async. Calling
+	/// this method in a sync handler without `await`ing it will do nothing.
+	pub async fn start_command(&self, cmd: Command, set: EventSet) -> SupervisorId {
```

Comment:
Ah, this should take a `Vec<Command>`

Reply:
oh, right!! all the methods still follow the `{something}_command` naming scheme, or would it be better to change this to `{something}_commands`?

Reply:
let's keep it as is for this PR

Reply:
okie!! :))